app/scripts/lib/WalletFundsObtainedMonitor.ts

```diff
@@ -0,0 +1,197 @@
+      const filteredNotifications = notifications.filter(
+        (n) =>
+          n.type === TRIGGER_TYPES.ERC20_RECEIVED ||
+          n.type === TRIGGER_TYPES.ETH_RECEIVED,
+      );
+
+      if (filteredNotifications.length < 1) {
+        return;
+      }
+
+      // Use the last (oldest) notification
+      const { chain_id: chainId, data } = filteredNotifications.at(-1) as {
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        chain_id: number;
+        data: { token?: { usd?: string }; amount?: { usd?: string } };
+      };
+
+      // ERC20 transfers have `token` object, native transfers have `amount` object
+      const amountUsd = data?.token?.usd || data?.amount?.usd;
+
+      if (chainId && amountUsd) {
+        this.#messenger.call(
+          'MetaMetricsController:trackEvent',
+          getWalletFundsObtainedEventProperties({
+            chainId,
+            amountUsd,
+          }),
+        );
+
+        this.#messenger.call(
+          'AppStateController:setCanTrackWalletFundsObtained',
+          false,
+        );
+
+        if (this.#handleWalletFundingNotification) {
+          this.#messenger.unsubscribe(
+            'NotificationServicesController:notificationsListUpdated',
+            this.#handleWalletFundingNotification,
+          );
+          this.#handleWalletFundingNotification = null;
+        }
+      }
+    };
+  }
+
+  /**
+   * Sets up monitoring to detect and track when a non-imported wallet first receives funds
+   * via ERC20 or ETH received events from the notification service.
+   */
+  setupMonitoring(): void {
+    // Avoid setting up the listener multiple times
+    if (this.#listenerSetup) {
+      return;
+    }
+
+    // Only target created wallets (not imported or restored)
+    const onboardingState = this.#messenger.call(
+      'OnboardingController:getState',
+    );
+    const { firstTimeFlowType } = onboardingState;
+
+    if (
+      firstTimeFlowType !== FirstTimeFlowType.create &&
+      firstTimeFlowType !== FirstTimeFlowType.socialCreate
+    ) {
+      this.#messenger.call(
+        'AppStateController:setCanTrackWalletFundsObtained',
+        false,
+      );
+      return;
+    }
+
+    // Only target wallets with notifications enabled
+    const notificationServicesState = this.#messenger.call(
+      'NotificationServicesController:getState',
+    );
+
+    if (!notificationServicesState.isNotificationServicesEnabled) {
+      this.#messenger.call(
+        'AppStateController:setCanTrackWalletFundsObtained',
+        false,
+      );
+      return;
+    }
+
+    if (this.#hasExistingFunds()) {
+      this.#messenger.call(
+        'AppStateController:setCanTrackWalletFundsObtained',
+        false,
+      );
+    } else {
+      this.#handleWalletFundingNotification =
+        this.#createWalletFundingNotificationHandler();
+      this.#messenger.subscribe(
+        'NotificationServicesController:notificationsListUpdated',
```

Comment:
Yep this should give you the dynamic updates on extension 👍🏾 

IIRC, in the near future, we can replace this with the `AccountActivityService:balanceUpdated` event - this should behave similar as is uses the new websocket subscriptions for balance updates (under an LD feature flag)
 
 https://github.com/MetaMask/metamask-extension/blob/e3f937bbb7390f527f0761fd76ad59e0305e5fca/app/scripts/controller-init/messengers/token-balances-controller-messenger.ts#L98

Reply:
Noted! One thing that sold the notifications API to us was the fact that it contains transaction info. Do you know if the AccountActivityService state would also contain this info or would we still have to source it elsewhere to determine the specific transaction the balance update is related to? @Prithpal-Sooriya